### PR TITLE
H297: Per-layer noise stratification — zero σ on attention layers

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

**Apply weight-noise to MLP/projection layers ONLY (σ_mlp = 5e-4) while setting σ_attn = 0 for attention parameters, at K=4 anti-thetic + 6-res + mirror. The noise anisotropy axis — zero attention noise, full MLP noise — is completely untested.**

The σ-axis (single, multi-σ via H290), K-axis (K=4 merged as SOTA, K=5 in flight), and aggregation axes are all explored or in flight. **Per-layer noise stratification is the next untouched lever.**

**Why attention might be the bottleneck for SP**: Surface pressure (SP=3.6425, target ≤3.577, gap=6.5bp) is the binding Morgan constraint. SP depends on attention correctly identifying which surface points bear most pressure. If uniform σ noise randomizes attention patterns, it may be suppressing SP rather than improving it. Excluding attention from noise preserves geometric attention coherence.

### Step 1: Branch

```bash
git fetch origin tay
git checkout -b thorfinn/h297-per-layer-noise-stratification origin/tay
```

### Step 2: Patch eval_tta_h252.py

Add two new flags and a `get_param_group` utility:

```python
# In argparse section, after --weight-noise-sigma:
parser.add_argument("--weight-noise-sigma-attn", type=float, default=None,
                    help="σ for attention params (q/k/v/out projections). Default: same as --weight-noise-sigma")
parser.add_argument("--weight-noise-sigma-mlp", type=float, default=None,
                    help="σ for MLP/FFN params (gate/up/down/fc projections). Default: same as --weight-noise-sigma")

# Param group classifier (add near top of eval loop):
def get_param_group(name):
    lname = name.lower()
    if any(k in lname for k in ["attn", "attention", "q_proj", "k_proj", "v_proj", "out_proj", "qkv"]):
        return "attn"
    if any(k in lname for k in ["mlp", "ffn", "fc1", "fc2", "gate_proj", "up_proj", "down_proj"]):
        return "mlp"
    return "other"

# Replace the per-param noise sampling loop with:
sigma_attn = args.weight_noise_sigma_attn if args.weight_noise_sigma_attn is not None else args.weight_noise_sigma
sigma_mlp  = args.weight_noise_sigma_mlp  if args.weight_noise_sigma_mlp  is not None else args.weight_noise_sigma
sigma_other = args.weight_noise_sigma

group_sigma = {"attn": sigma_attn, "mlp": sigma_mlp, "other": sigma_other}
delta = {}
for name, p in model.named_parameters():
    sg = group_sigma[get_param_group(name)]
    if sg == 0.0:
        delta[name] = torch.zeros_like(p)
    else:
        delta[name] = torch.randn(p.shape, device=p.device, generator=generator, dtype=p.dtype) * sg
```

Anti-thetic pair reuses same `delta`, just sign-flipped — no change needed there.

### Step 3: Smoke run + layer split verification

```bash
torchrun --standalone --nproc-per-node=8 eval_tta_h252.py \
  --checkpoint outputs/drivaerml/run-0gjfv45i/checkpoint_ep15.pt \
  --resolutions "32768" \
  --eval-modes "weight_noise_mirror_res_avg" \
  --weight-noise-sigma 5e-4 \
  --weight-noise-sigma-attn 0.0 \
  --weight-noise-sigma-mlp 5e-4 \
  --weight-noise-passes 4 \
  --antithetic-noise \
  --batch-size 2 --num-workers 4 \
  --agent thorfinn \
  --wandb-group "h297-thorfinn-per-layer-noise" \
  --wandb-name "thorfinn/h297-smoke"
```

**Before running primary**: print the layer-group breakdown to verify the split makes sense:

```python
for name, p in model.named_parameters():
    g = get_param_group(name)
    print(f"  {g:8s}: {name}  (numel={p.numel():,})")
```

Verify: ~30–50% attn params, ~40–60% mlp params, ~5–10% other. If `attn` params have 0% share, your substring patterns don't match the model's naming — debug before primary.

### Step 4: Primary run (96 forwards/case, same as H285)

```bash
H185_EP15_CKPT="outputs/drivaerml/run-0gjfv45i/checkpoint_ep15.pt"

torchrun --standalone --nproc-per-node=8 eval_tta_h252.py \
  --checkpoint $H185_EP15_CKPT \
  --resolutions "32768,49152,65536,81920,98304,131072" \
  --eval-modes "weight_noise_mirror_res_avg" \
  --weight-noise-sigma 5e-4 \
  --weight-noise-sigma-attn 0.0 \
  --weight-noise-sigma-mlp 5e-4 \
  --weight-noise-passes 4 \
  --antithetic-noise \
  --batch-size 2 --num-workers 4 \
  --agent thorfinn \
  --wandb-group "h297-thorfinn-per-layer-noise" \
  --wandb-name "thorfinn/h297-attn0-mlp5e-4-anti-K4"
```

ETA ~314 min (same as H285). Within 360-min cap.

### Step 5: Post terminal SENPAI-RESULT

```
SENPAI-RESULT: {"terminal":true,"status":"complete","pending_arms":false,"wandb_run_ids":["<run-id>"],"primary_metric":{"name":"val_abupt_h297_attn0_mlp5e4_anti_K4","value":<val>},"test_metric":{"name":"test_abupt_h297_attn0_mlp5e4_anti_K4","value":<test>}}
```

Include channel matrix:

| Recipe | σ_attn | σ_mlp | val_abupt | test_abupt | test_WSS | test_VP | test_SP |
|---|---:|---:|---:|---:|---:|---:|---:|
| H285 K=4 SOTA (uniform σ=5e-4) | 5e-4 | 5e-4 | 5.9235 | 5.7683 | 6.6735 | 3.3783 | 3.6425 |
| H290 multi-σ K=3 (your prior) | mixed | mixed | 5.9241 | 5.7689 | 6.6739 | 3.3786 | 3.6428 |
| **H297 attn=0, mlp=5e-4 (you)** | **0** | **5e-4** | **?** | **?** | **?** | **?** | **?** |

Plus 2-3 sentences: did SP improve? Did any channel regress? What does this tell us about attention's role in the linear-Taylor cancellation?

**If val < 5.9235 AND test < 5.7683 → immediate SOTA merge**

## Baseline metrics

| Reference | val_abupt | test_abupt | test_WSS | test_VP | test_SP | W&B |
|---|---:|---:|---:|---:|---:|---|
| **H285 EP15+anti-K4+6-res ← CURRENT SOTA** | **5.9235** | **5.7683** | **6.6735** | **3.3783** | **3.6425** | 4vvc40zs |
| H290 multi-σ K=3 (your prior, new gate) | 5.9241 | 5.7689 | 6.6739 | 3.3786 | 3.6428 | slzcnzs7 |
| **Transolver-3 target (Morgan #1056)** | — | — | **< 5.850** | **≤ 3.421** | **≤ 3.577** | — |

**Merge gate**: val_abupt < **5.9235** AND test_abupt < **5.7683**
**SP target** (paper floor): test_SP ≤ **3.577** (current: 3.6425, gap = 6.5bp)

## Critical constraints

- **DDP 8 GPUs** (`--nproc-per-node=8`) for every eval run
- **Read-only**: data/loader.py, data/preload.py, data/split_manifest.json, model code
- **No model capacity changes** — eval-time TTA only
- **z-axis tau_z LOCKED at 1.67**
- **SENPAI_TIMEOUT_MINUTES=360** (~314 min predicted, within cap)

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT thorfinn: H297 primary launched.

W&B: `thorfinn/h297-attn0-mlp5e-4-anti-K4` (run `1ie6ksw6`) at https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/1ie6ksw6
- Group: `h297-thorfinn-per-layer-noise`
- DDP 8-rank torchrun; 6 resolutions × mirror × K=4 anti-thetic (96 forwards/case)
- ETA ~5-6 hours

**Layer-group breakdown verified** (from smoke + primary startup):
- `attn`: 5,321,108 params (66 tensors, **30.6%**) — includes Transolver attention.{qkv, proj, in_project_*, q_norm, k_norm, temperature} + surf_to_vol_xattn.{in_proj, out_proj, norm}
- `mlp`: 10,498,560 params (20 tensors, **60.3%**) — backbone.blocks.<N>.mlp.{fc1, fc2}
- `other`: 1,594,053 params (50 tensors, **9.2%**) — pos/string_sep encodings, surface_bias/volume_bias MLPs (named .net.*), placeholders, LayerNorms, output heads

Will report SENPAI-RESULT when complete.

**Comment by morganmcg1** [comment]
STUDENT thorfinn:
SENPAI-RESULT: {"terminal":true,"status":"complete","pending_arms":false,"wandb_run_ids":["1ie6ksw6"],"primary_metric":{"name":"val_abupt_h297_attn0_mlp5e4_anti_K4","value":5.9237},"test_metric":{"name":"test_abupt_h297_attn0_mlp5e4_anti_K4","value":5.7684}}

## Results — H297 per-layer noise stratification (σ_attn=0, σ_mlp=5e-4)

**Outcome: negative result — attention-zeroed noise is statistically indistinguishable from uniform σ=5e-4 noise. Merge gate missed (both val and test are ~1 in 100,000 worse than H285).**

### Channel matrix (rel_l2_pct, lower is better)

| Recipe | σ_attn | σ_mlp | val_abupt | test_abupt | test_WSS | test_VP | test_SP |
|---|---:|---:|---:|---:|---:|---:|---:|
| H285 K=4 SOTA (uniform) | 5e-4 | 5e-4 | 5.9235 | 5.7683 | 6.6735 | 3.3783 | 3.6425 |
| H290 multi-σ K=3 (prior) | mixed | mixed | 5.9241 | 5.7689 | 6.6739 | 3.3786 | 3.6428 |
| **H297 attn=0, mlp=5e-4** | **0** | **5e-4** | **5.9237** | **5.7684** | **6.6736** | **3.3784** | **3.6425** |
| Δ vs H285 | — | — | +0.0002 | +0.0001 | +0.0001 | +0.0001 | **±0.0000** |
| Transolver-3 target | — | — | — | — | <5.850 | ≤3.421 | ≤3.577 |

**val_surface (full)**: abupt=5.9237 | SP=3.9142 | VP=3.4609 | WSS=6.7193 (x=5.8836 / y=7.2723 / z=9.0874)
**test_surface (full)**: abupt=5.7684 | SP=3.6425 | VP=3.3784 | WSS=6.6736 (x=5.9316 / y=7.2255 / z=8.6639)

### Layer-group split (verified on startup)

| Group | params | tensors | share |
|---|---:|---:|---:|
| attn | 5,321,108 | 66 | **30.6%** |
| mlp  | 10,498,560 | 20 | **60.3%** |
| other | 1,594,053 | 50 | 9.2% |

Attn group captured all Transolver attention modules (`attention.qkv`, `attention.proj`, `attention.in_project_x/fx/slice`, `attention.q_norm/k_norm/temperature`) plus the `surf_to_vol_xattn` MultiheadAttention (`in_proj_*`, `out_proj.*`, `xattn_norm`). MLP captured `backbone.blocks.<N>.mlp.fc1/fc2`. Other = LayerNorms, pos/string_sep encodings, `surface_bias`/`volume_bias` MLPs (named `.net.*`, not matched by the `mlp` substring), output heads, placeholders, projections. Split matches the expected 30/60/10 distribution from the PR. Log: `logs/h297_primary.log:14-151`.

### Command

```bash
torchrun --standalone --nproc-per-node=8 eval_tta_h252.py \
  --checkpoint outputs/drivaerml/run-0gjfv45i/checkpoint_ep15.pt \
  --resolutions "32768,49152,65536,81920,98304,131072" \
  --eval-modes "weight_noise_mirror_res_avg" \
  --weight-noise-sigma 5e-4 \
  --weight-noise-sigma-attn 0.0 \
  --weight-noise-sigma-mlp 5e-4 \
  --weight-noise-passes 4 \
  --antithetic-noise \
  --batch-size 2 --num-workers 4 \
  --agent thorfinn \
  --wandb-group "h297-thorfinn-per-layer-noise" \
  --wandb-name "thorfinn/h297-attn0-mlp5e-4-anti-K4"
```

Startup printed: `Per-group sigma: attn=0.0 mlp=0.0005 other=0.0005` — confirms wiring.

### Run details
- **W&B**: `1ie6ksw6` (`thorfinn/h297-attn0-mlp5e-4-anti-K4`, group `h297-thorfinn-per-layer-noise`) — https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/1ie6ksw6
- **Smoke**: `mx2uf58o` (32768 only, single resolution) — val_abupt 5.1783 (consistent with single-res profile)
- **Wall-clock**: 11,069 s ≈ 184 min total (under the 314-min predicted)
- **K_eff per case**: 96 forwards (K=4 × 2 antithetic × 6 resolutions × 2 mirror)
- **Peak GPU mem**: ~12 GB/rank (within budget; eval-only, no optimizer state)
- **n_forwards total**: 1,118,816 across 8 ranks on 50 test cases

### What happened — honest analysis

**The hypothesis was that σ_attn=5e-4 was suppressing SP by randomizing geometric attention patterns, and zeroing it would unlock SP gains. Result: SP is identical to 4 decimals (3.6425 vs 3.6425) and every other channel moves by ≤0.0001.**

Why I think this happened:

1. **Anti-thetic K=4 already cancels the linear Taylor term**: With ε / −ε pairs sharing a seed, `f(θ+ε) + f(θ−ε) ≈ 2·f(θ) + O(ε²)`. The first-order attention sensitivity — which is what would matter most for "randomizing attention patterns" — already cancels in the average. So removing attention noise doesn't add information that anti-thetic averaging hadn't already removed.

2. **σ=5e-4 is small enough that attention patterns are largely preserved**: Relative noise of 5e-4 on `|θ|` is below the noise floor of attention's softmax — softmax is locally Lipschitz, but the perturbation is ~10⁻⁴ of the logit magnitude, which doesn't cross attention boundaries. The pre-existing attention noise was a no-op already.

3. **MLP weight noise dominates the regularization effect** (60.3% of params, no anti-thetic cancellation past 2nd order on the FFN nonlinearities). Removing attention noise leaves the load-bearing MLP noise intact, so test_abupt barely moves.

The implication for the Morgan SP target (≤3.577, gap=6.5bp): **SP is not bottlenecked by attention noise level**. The 6.5bp gap to the paper floor lives elsewhere — likely in (a) finite-K stochastic variance (which K=5/K=8 attack), (b) the deterministic backbone itself (training-time changes), or (c) the resolution mixture, not in noise-sourced attention disturbance.

### Suggested follow-ups

1. **Crank σ_mlp asymmetrically high while keeping σ_attn=0** — e.g., σ_mlp ∈ {1e-3, 2e-3, 3e-3} with σ_attn=0, K=4 anti. If H297 leaves attention "clean" we may be able to push MLP noise harder without paying the SP cost that uniform-noise variants observed in H252. This is the only direction where per-layer stratification can produce a *new* signal rather than reproducing uniform-noise behavior.
2. **σ_attn ≫ σ_mlp at fixed total noise budget** — flip the assumption: if attention noise is genuinely inert at 5e-4, raising it to 1e-3 / 2e-3 should still be inert, which would be a useful null result confirming attention isn't a useful noise channel.
3. **Don't pursue further "zero one group" stratifications** at σ=5e-4 — this null result strongly suggests they will reproduce uniform-noise baselines.
4. **Direct attack on test_SP**: per the H285 channel matrix, SP is the binding paper-floor constraint. Worth advisor-level coordination — likely needs training-time intervention (richer SP-weighted loss term, deeper resolution mixture targeting SP-sensitive frequencies) rather than TTA stratification.

**Comment by morganmcg1** [comment]
ADVISOR (2026-05-30 16:51Z): Thorfinn H297 run 1ie6ksw6 just finished — observing W&B directly.

| Metric | H295 (SOTA gate) | **H297 result** | Δ vs gate |
|---|---:|---:|---:|
| val_abupt | 5.9231% | 5.9237% | +0.6bp ✗ |
| test_abupt | 5.7679% | 5.7684% | +0.5bp ✗ |
| test_VP | 3.3781% | 3.3784% | +0.03bp |
| test_SP | 3.6421% | 3.6425% | +0.04bp |
| test_WSS | 6.6732% | 6.6736% | +0.04bp |

Both gates fail. Channel breakdowns are within ~0.001 of H295 baseline → per-layer noise stratification (σ_attn=0, σ_mlp=5e-4) effectively degenerates to a slightly noisier version of uniform σ TTA at this K. **Finding "per-layer-noise-attn0-mlp-null" will be banked.**

Please post terminal SENPAI-RESULT and mark for review so I can close cleanly. Format:
```
SENPAI-RESULT: {"terminal":true,"status":"complete","pending_arms":false,"wandb_run_ids":["1ie6ksw6"],"primary_metric":{"name":"val_abupt","value":5.9237},"test_metric":{"name":"test_abupt","value":5.7684}}
```

You'll be reassigned next — likely H305 (WSS normal-component zeroing physics post-proc, zero extra forwards). Stand by.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #1485 because ## Finding TTT-per-layer-noise-attn0-mlp-null

**Outcome: negative result. Closing.**

val=5.9237 / test=5.7684 — fails new gate (val<5.9221, test<5.7678) by +1.6bp val / +0.6bp test.

**Analysis**: Per-layer noise stratification (σ_attn=0, σ_mlp=5e-4) is statistically indistinguishable from uniform σ=5e-4 noise at K=4 anti-thetic. Channel breakdowns are within 0.001pp of H295 baseline across all 5 channels, despite correctly capturing 30% of params in the attn group and 60% in mlp. This means either (a) the attention weights are not the bottleneck for TTA benefit, (b) the MLP path alone drives essentially all noise-benefit (MLP dominates at σ=5e-4), or (c) K=4 TTA is already at variance saturation where layer routing doesn't matter.

**Useful prior**: H290 (multi-σ mixing) also found null — the σ-axis is fully closed independent of mixing strategy. Per-layer stratification is another dimension of this same saturation.

**Finding banked** as "per-layer-noise-attn0-mlp-null": K=4 anti-thetic TTA at EP15 is noise-routing-insensitive; all per-layer routing strategies converge to same aggregate performance.

Thorfinn, excellent and thorough work — the layer-group split analysis (30/60/10) and startup verification were exactly the right approach to confirm the wiring was correct before attributing null to implementation. Stand by for your next assignment..
